http/src/main/scala/org/http4s/blaze/http/BodyReader.scala

```diff
@@ -0,0 +1,125 @@
+package org.http4s.blaze.http
+
+import java.nio.ByteBuffer
+
+import org.http4s.blaze.util.{BufferTools, Execution}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success}
+
+/** Representation of a HTTP message body
+  *
+  * @note The release of resources must be idempotent, meaning that `discard()` may
+  *       be called after complete consumption of the body and it may also be called
+  *       numerous times.
+  */
+trait BodyReader {
+
+  /** Throw away this `MessageBody` */
+  def discard(): Unit
+
+  /** Get a `Future` which may contain message body data.
+    *
+    * If no data remains, the `ByteBuffer` will be empty as defined by `ByteBuffer.hasRemaining()`
+    */
+  def apply(): Future[ByteBuffer]
+
+  /** Examine whether the `MessageBody` may yield additional data.
+    *
+    * This may be a result of being discarded, failure, or deletion of the data stream.
+    *
+    * Because `MessageBody` is async it is not, in general, possible to definitively determine
+    * if more data remains in the stream. Therefore, the contract of this method is that a return
+    * value of `true` guarantees that no more data can be obtained from this `MessageBody`, but a
+    * return value of `false` does not guarantee more data.
+    */
+  def isEmpty: Boolean
```

Comment:
I'm not real thrilled with the name in light of that comment.  `isExhausted`, maybe?